test/src/test/java/org/corfudb/integration/StreamingIT.java

```diff
@@ -73,27 +74,47 @@ public void loadProperties() {
      * A StreamListener implementation to be used in the tests.
      * This listener accumulates all updates streamed to it into a linked list that
      * can then be used to verify.
+     * This listener can throw an exception on 'n' number of updates, in order to
+     * trigger the onError(Throwable t) and consequently, automatically unsubscribe.
```

Comment:
What type of errors would cause the listener to briefly de-register then register again ?

`onError` should be used to notify the listen of corfu related errors, usually shutdown/trim exceptions. In those two cases, the listener will not resume from a know sequence number, but it will re-register from the current time. 

It's not clear how exposing this sequence number helps reconcile issues for the aforementioned type of errors.

Reply:
The scenario that they were discussing initially was the case of any Corfu failure from which they'd be able to recover (I'm trying to think what this could be, I can only think of some race condition or network connection, cause they also talked about a poison pill, but in any case we won't be able to recover from that). The case that I do see more relevant is any sort of processing exception (business logic in the onNext, which we can't prevent from occurring) would also cause onError to be triggered and consequently unsubscribe. They want a means to be able to subscribe from the last synced point, rather than having to do a full sync.

Reply:
I think the intention of SubscriptionException was to signal corfu related exceptions, most of which won't allow the consumer to resume from the last successful point. For example, TrimmedException. Network exceptions and other related corfu exceptions shouldn't be passed to the clients since they can't really handle them (ex NetworkException).

Business logic should handle their own exceptions. I don't see the benefit of them not handling an exception, then we catch it and return it to them wrapped in another exception.

One possible use case is producers and consumers trying to coordinate via these subscriptions. Where a producer might write a command and the consumers would unsubscribe once they receive that command via onNext. I'm just trying to see if its possible to hide the sequence/epoch from clients.

Reply:
Yes, a TrimmedException is undoubtedly an unrecoverable error. 

As per business logic handling their own exceptions, I agree, but that is not the case now, we still need to make changes to our code to be resilient to that. We currently catch those exceptions and enter onError state (that is actually how I simulated a failure in the test). 

Ok, I see but why do we want to hide it if anyways we ask for it in the subscription API?

Reply:
Reattaching a stream onError() actually leads to a chain failure scenario as we had seen in one of the early clients. Hope the timestamp is not meant to be used purely for onError() based re-subscription since the right thing as @Maithem mentioned is for the client to open a snapshot before continuing.
Exposing the timestamp is fine, but perhaps we need to clarify the intention in the ask?